### PR TITLE
Created tables for Operator, Client, Tool

### PR DESCRIPTION
**Pull request by abjjabjj**
Using FastAPI to create REST API for persistent operator interaction

Created Operator, Client, and Tool Table:

```
Operator:
  instructions: str
  title: str(32)
  
  clients: list[Client]
  tools: list[Tool]
  
 Client:
   client: str(32)
   temperature: float
   model: str(32)
   operator_id: foreign key
   operator: Operator
   
 Tool:
  operator_id: foreign key
  operator: Operator
```

Create and Read work on operator, but Update and Delete do not. CRUD not yet implemented for other objects.

### CODE REVIEW

File: concrete/abstract.py [hunk 1]
```diff
@@ -109,7 +109,7 @@ def _delay(
 class AbstractOperator(metaclass=MetaAbstractOperator):
 
     # TODO replace OpenAIClient with GenericClient
-    def __init__(self, clients: dict[str, OpenAIClient], tools: Optional[List[MetaTool]] = None):
+    def __init__(self, clients: dict[str, OpenAIClient], tools: list[MetaTool] | None = None):
```
Comment (MichaelDeng03) [inline concrete/abstract.py:112]: We should establish a standard for typing. | None should be equivalent to Optional.

File: pyproject.toml [hunk 2]
```diff
@@ -44,13 +43,21 @@ uvicorn = "==0.30.4"
 fastapi = "==0.111.1"
 
 
+[tool.poetry.group.webapp-api.dependencies]
+sqlalchemy = "^2.0.34"
```
Comment (MichaelDeng03) [inline pyproject.toml:47]: Lock 🙏

File: webapp/api/crud.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
+from uuid import UUID
+
+from sqlalchemy import delete, select, update
+from sqlalchemy.orm import Session
+
+from . import models, schemas
+
+
+def get_operator(db: Session, operator_id: UUID) -> models.Operator | None:
```
Comment (MichaelDeng03) [inline webapp/api/crud.py:9]: Naming session db might be a little ambiguous. Db could refer to an engine, connection, and probably others

File: webapp/api/models.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
+from uuid import UUID, uuid4
```
Comment (MichaelDeng03) [inline webapp/api/models.py:1]: Maybe I'm missing it, but are you creating tables? Table creation is idempotent so maybe throw it in init?

File: webapp/api/crud.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+# TODO: Doesn't do anything
+def update_operator(db: Session, operator_id: UUID, operator: schemas.OperatorUpdate) -> models.Operator | None:
+    print(operator.model_dump(exclude_none=True))
+    stmt = (
+        update(models.Operator)
+        .where(models.Operator.id == operator_id)
+        .values(operator.model_dump(exclude_none=True))
+        .returning(models.Operator)
+    )
```
Comment (MichaelDeng03) [inline webapp/api/crud.py:35]: Toss in a commit

File: pyproject.toml (hunk 2 above)
Comment (kgang) [reply pyproject.toml:47]: im thinking this should be a mian dep also

File: concrete/abstract.py (hunk 1 above)
Comment (abjjabjj) [reply concrete/abstract.py:112]: I prefer using built-in typing when possible to reduce the number of imports

Comment (kgang) [reply concrete/abstract.py:112]: I like `| None` more personally but i think if something is style related it shouldn't be a part of PR reviews

File: pyproject.toml (hunk 2 above)
Comment (abjjabjj) [reply pyproject.toml:47]: We should reorganize all of this

Comment (kgang) [reply pyproject.toml:47]: whats wrong with it [](edited-with-Axolo)

Comment (kgang) [reply pyproject.toml:47]: feel free to reorganize [](edited-with-Axolo)

File: webapp/api/models.py (hunk 4 above)
Comment (abjjabjj) [reply webapp/api/models.py:1]: table creation happens in `api/server.py:11`, but we can certainly move it elsewhere

File: pyproject.toml (hunk 2 above)
Comment (abjjabjj) [reply pyproject.toml:47]: well this for one but we have redundant dep for our `webapp-*` groups that can be put together into a single `webapp` group [](edited-with-Axolo)

Comment (MichaelDeng03) [reply pyproject.toml:47]: Agreed [](edited-with-Axolo)

File: webapp/api/crud.py (hunk 3 above)
Comment (kgang) [reply webapp/api/crud.py:9]: I think if this is how the docs have it written its good enough [](edited-with-Axolo)

Comment (abjjabjj) [reply webapp/api/crud.py:9]: what do you prefer? it's already typed `Session` but if you like I can rename to `session`

File: webapp/api/crud.py (hunk 5 above)
Comment (kgang) [reply webapp/api/crud.py:35]: does the db/session object hold onto this in memory waiting for a `db.commit()` or will it be reset/refreshed next time a transaction is started?

Comment (MichaelDeng03) [reply webapp/api/crud.py:35]: Won't be reset but the next transaction won't know if it until it's flushed [](edited-with-Axolo)

Comment (kgang) [reply webapp/api/crud.py:35]: lol i dont know what that means [](edited-with-Axolo)

Comment (MichaelDeng03) [reply webapp/api/crud.py:35]: Flushing marks it as a change, and commiting it writes it to disk [](edited-with-Axolo)

Comment (kgang) [reply webapp/api/crud.py:35]: what does won't be reset mean [](edited-with-Axolo)

Comment (MichaelDeng03) [reply webapp/api/crud.py:35]: It won't be lost, it'll be stored in the session object [](edited-with-Axolo)

Comment (MichaelDeng03) [reply webapp/api/crud.py:35]: Probably flagged as dirty [](edited-with-Axolo)

Comment (MichaelDeng03) [reply webapp/api/crud.py:35]: But it won't be used for anything [](edited-with-Axolo)

Comment (kgang) [reply webapp/api/crud.py:35]: I see [](edited-with-Axolo)

Comment (kgang) [reply webapp/api/crud.py:35]: it won't be applied either in `db.commit`you're saying [](edited-with-Axolo)

Comment (MichaelDeng03) [reply webapp/api/crud.py:35]: No it will be applied in commit [](edited-with-Axolo)

Comment (MichaelDeng03) [reply webapp/api/crud.py:35]: Flushing will apply it, and commits always flush [](edited-with-Axolo)

Comment (kgang) [reply webapp/api/crud.py:35]: so the answer to my original question is that the first statement is correct? [](edited-with-Axolo)

Comment (MichaelDeng03) [reply webapp/api/crud.py:35]: Oh yes lol [](edited-with-Axolo)

Comment (MichaelDeng03) [reply webapp/api/crud.py:35]: It'll be queued for application, but other operations won't know of it until it's actually flushed or committed [](edited-with-Axolo)

Comment (MichaelDeng03) [reply webapp/api/crud.py:35]: So you could select after and insert, and the select won't find it [](edited-with-Axolo)

Comment (MichaelDeng03) [reply webapp/api/crud.py:35]: Autoflush does exist though [](edited-with-Axolo)

Comment (kgang) [reply webapp/api/crud.py:35]: I understand now thank. Seems inaccurate to say these functions do nothing then @abjjabjj [](edited-with-Axolo)

File: pyproject.toml (hunk 2 above)
Comment (abjjabjj) [reply pyproject.toml:47]: if we make sqlalchemy as a main dep then we can move the db files into main then agreed? [](edited-with-Axolo)

Comment (kgang) [reply pyproject.toml:47]: yeah we should [](edited-with-Axolo)

File: webapp/api/crud.py (hunk 5 above)
Comment (abjjabjj) [reply webapp/api/crud.py:35]: good catch @MichaelDeng03 , and not wholly inaccurate @kgang . at least not until commit is made to db

Comment (kgang) [reply webapp/api/crud.py:35]: OK ig technically technically is does nothing but adds extra random flavor to the next transaction :shrug: [](edited-with-Axolo)

Comment (MichaelDeng03) [reply webapp/api/crud.py:35]: A functional programmers wet dream [](edited-with-Axolo)

File: pyproject.toml [hunk 6]
```diff
@@ -14,13 +14,16 @@ openai = "==1.40.5"
 astroid = "==3.2.3"
 astor = "==0.8.1"
 celery = "==5.4.0"
+sqlalchemy = "==2.0.34"
+alembic = "==1.13.2"
```
Comment (kgang) [inline pyproject.toml:18]: is alembic used in this PR?

**Review by kgang — COMMENTED** [review]
LGTM but I'll defer to Michael

File: webapp/api/server.py [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+from typing import Annotated
+from uuid import UUID
+
+from fastapi import Depends, FastAPI, HTTPException
+from sqlalchemy.orm import Session
+
+from . import crud, models, schemas
+from .database import SessionLocal, engine
+
+models.Base.metadata.create_all(bind=engine)
```
Comment (kgang) [inline webapp/api/server.py:11]: Can you add a comment describing what the behavior is if the db already exists?

File: pyproject.toml (hunk 6 above)
Comment (abjjabjj) [reply pyproject.toml:18]: it is not, but this is just the intermediary PR anyways [](edited-with-Axolo)

Comment (kgang) [reply pyproject.toml:18]: Ok [](edited-with-Axolo)

Comment (abjjabjj) [reply pyproject.toml:18]: iirc you just wanted to get the table definitions merged in [](edited-with-Axolo)

Comment (kgang) [reply pyproject.toml:18]: yeah just checking [](edited-with-Axolo)

File: webapp/api/server.py [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+
+def get_db():
+    db = SessionLocal()
+    try:
+        yield db
+    finally:
+        db.close()
+
+
+DbDep = Annotated[Session, Depends(get_db)]
```
Comment (MichaelDeng03) [inline webapp/api/server.py:29]: very cool

File: webapp/api/models.py [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+
+
+class Base(DeclarativeBase):
+    @declared_attr.directive
+    def __tablename__(cls) -> str:
+        return cls.__name__.lower()
+
+    id: Mapped[UUID] = mapped_column(primary_key=True, default=uuid4)
+
+    def __repr__(self) -> str:
```
Comment (MichaelDeng03) [inline webapp/api/models.py:20]: https://docs.sqlalchemy.org/en/20/orm/mapping_api.html#sqlalchemy.orm.class_mapper
+
https://docs.sqlalchemy.org/en/20/orm/mapping_api.html#sqlalchemy.orm.Mapper.columns
offers more flexibility than .columns in the context of ORM constructs. As far as I can tell, .c won't display inherited columns (like id), but Mapping.columns will.

File: webapp/api/models.py [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+        attrs_str = ", ".join([f"{c}={self.__table__.columns[c]!r}" for c in self.__table__.columns])
+        return f"{self.__class__.__name__}({attrs_str})"
+
+
+class Operator(Base):
+    instructions: Mapped[str]
+    title: Mapped[str] = mapped_column(String(32))
+
+    clients: Mapped[list["Client"]] = relationship(
```
Comment (MichaelDeng03) [inline webapp/api/models.py:29]: Wouldn't this delete its associated Client? My understanding is that a Client can have multiple Operators.

File: webapp/api/models.py [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+
+    id: Mapped[UUID] = mapped_column(primary_key=True, default=uuid4)
+
+    def __repr__(self) -> str:
+        attrs_str = ", ".join([f"{c}={self.__table__.columns[c]!r}" for c in self.__table__.columns])
+        return f"{self.__class__.__name__}({attrs_str})"
+
+
+class Operator(Base):
+    instructions: Mapped[str]
```
[the remark this answers is not in the document]
Comment (abjjabjj) [reply webapp/api/models.py:26]: unless you think we need restrictions on length or other attributes, this is legal

File: webapp/api/models.py (hunk 10 above)
Comment (abjjabjj) [reply webapp/api/models.py:29]: it would. do we need to share clients between operators?

Comment (MichaelDeng03) [reply webapp/api/models.py:29]: Probably not, what's the current behavior? [](edited-with-Axolo)

Comment (abjjabjj) [reply webapp/api/models.py:29]: currently the celery task doesn't even pass around clients, it recreates them using the pydantic model. My point being that restricting clients to a single operator like this reduces errors and complexity in our codebase [](edited-with-Axolo)

Comment (abjjabjj) [reply webapp/api/models.py:29]: if we have a legitimate need to allow clients to have multiple operators though that's a different story [](edited-with-Axolo)

Comment (MichaelDeng03) [reply webapp/api/models.py:29]: :+1: lgtm then [](edited-with-Axolo)

**Review by MichaelDeng03 — APPROVED** [review]

**Comment by abjjabjj** [comment]
tyty

**Pull request merged**